### PR TITLE
feat(undici): request rewamp

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

**Comment by ronag** [comment]
@metcoder95 this was actually what I wanted you to look at.

File: undici/index.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    dispatch = interceptors.log(dispatch)
+    dispatch = interceptors.abort(dispatch)
+    dispatch = interceptors.signal(dispatch)
+    dispatch = interceptors.retryStatus(dispatch)
+    dispatch = interceptors.retryBody(dispatch)
+    dispatch = interceptors.redirect(dispatch)
+    dispatch = interceptors.contentLength(dispatch)
+    dispatch = interceptors.contentMD5(dispatch)
+    dispatch = interceptors.dump(dispatch)
```
Comment (metcoder95) [inline undici/index.js:105]: It's a really smart way of chaining dispatcher. I'm starting to guess if we should either document this or improve the DX for allowing chain dispatchers 🤔

File: undici/interceptor/dump.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    this.handler.onComplete([])
+    this.handler = null
+    return true
+  }
+
+  onData(chunk) {
+    this.position += chunk.length
+    if (this.position < 128 * 1024) {
+      return true
+    }
```
Comment (metcoder95) [inline undici/interceptor/dump.js:23]: Just to confirm my understanding, the point is just throw away the body by reading just 128kb?

File: undici/interceptor/retry-body.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    if (this.count >= 0 && this.count++ < this.opts.retry.count && !isDisturbed(this.opts.body)) {
+      const start = this.range.start + this.position
+      const end = this.range.end ? this.range.end : null
+
+      this.opts = {
+        ...this.opts,
+        headers: {
+          ...this.opts.headers,
+          range: `bytes=${start}-${end ?? ''}`,
+        },
```
Comment (metcoder95) [inline undici/interceptor/retry-body.js:93]: I know this is pretty specific to your use case, but here it should be about extensibility, userland can decide whether or not depends on the range or not; and important to allow default retries on idempotent method (e.g. `HEAD`, `GET`, `PUT`) and deny on others, unless userland submits a way to decide wether or not to retry 🤔

File: undici/interceptor/retry-body.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      }
+
+      this.timeout = setTimeout(
+        () => {
+          this.timeout = null
+          this.dispatch(this.opts, this)
+        },
+        Math.min(10e3, this.count * 1e3),
```
Comment (metcoder95) [inline undici/interceptor/retry-body.js:101]: I'm thinking of starting with something simpler as https://github.com/metcoder95/one-more-time/blob/4de4e460fabbcea3fc11d7daf158b9455bede2ae/lib/task.js#L145 which can go linear or exponential accordingly to a given factor, wdyt?

**Review by metcoder95 — COMMENTED** [review]
Nice work, I never thought of this approach for abstracting several implementations through layers of dispatchers. I'll do a second read later on or Tomorrow

File: undici/interceptor/dump.js (hunk 2 above)
Comment (ronag) [reply undici/interceptor/dump.js:23]: Yes. In order to keep the keep alive connection.

**Pull request merged**